clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -0,0 +1,177 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::trait_ref_of_method;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_errors::MultiSpan;
+use rustc_hir::intravisit::{walk_impl_item, walk_item, walk_param_bound, walk_ty, Visitor};
+use rustc_hir::{
+    GenericParamKind, Generics, ImplItem, ImplItemKind, Item, ItemKind, LifetimeParamKind, Ty, TyKind, WherePredicate,
+};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::hir::nested_filter;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{def_id::DefId, BytePos, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for type parameters in generics that are never used anywhere else.
+    ///
+    /// ### Why is this bad?
+    /// Functions cannot infer the value of unused type parameters; therefore, calling them
+    /// requires using a turbofish, which serves no purpose but to satisfy the compiler.
+    ///
+    /// ### Example
+    /// ```rust
+    /// // unused type parameters
+    /// fn unused_ty<T>(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn no_unused_ty(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    #[clippy::version = "1.67.0"]
+    pub EXTRA_UNUSED_TYPE_PARAMETERS,
+    complexity,
+    "unused type parameters in function definitions"
+}
+declare_lint_pass!(ExtraUnusedTypeParameters => [EXTRA_UNUSED_TYPE_PARAMETERS]);
+
+struct TypeWalker<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    map: FxHashMap<DefId, Span>,
+    bound_map: FxHashMap<DefId, Span>,
+    generics: &'tcx Generics<'tcx>,
+    some_params_used: bool,
+    has_non_ty_params: bool,
+}
+
+impl<'cx, 'tcx> TypeWalker<'cx, 'tcx> {
+    fn new(cx: &'cx LateContext<'tcx>, generics: &'tcx Generics<'tcx>) -> Self {
+        let mut has_non_ty_params = false;
+        let map = generics
+            .params
+            .iter()
+            .filter_map(|param| match &param.kind {
+                GenericParamKind::Type { .. } => Some((param.def_id.into(), param.span)),
+                non_ty => {
+                    if let GenericParamKind::Lifetime { kind } = non_ty {
+                        match kind {
+                            LifetimeParamKind::Elided => (),
+                            _ => has_non_ty_params = true,
+                        }
+                    }
+                    None
```

Comment:
Just realized this check is buggy; it doesn't account for constant parameters. Currently away from my PC, but I'll fix it tonight.